examples/pytorch/geniepath/ppi.py

```diff
@@ -0,0 +1,130 @@
+from dgl.data import PPIDataset
+from dgl.dataloading import GraphDataLoader
+from sklearn.metrics import f1_score
+
+from model import GeniePath, GeniePathLazy
+
+
+def evaluate(model, loss_fn, dataloader, device='cpu'):
+    loss = 0
+    f1 = 0
+    num_blocks = 0
+    for subgraph in dataloader:
+        subgraph = subgraph.to(device)
+        label = subgraph.ndata['label'].to(device)
+        feat = subgraph.ndata['feat']
+        logits = model(subgraph, feat)
+
+        # compute loss
+        loss += loss_fn(logits, label).item()
+        predict = np.where(logits.data.cpu().numpy() >= 0., 1, 0)
+        f1 += f1_score(label.cpu(), predict, average='micro')
+        num_blocks += 1
+
+    return f1 / num_blocks, loss / num_blocks
+
+
+def main(args):
+    # Step 1: Prepare graph data and retrieve train/validation/test index ============================= #
+    # Load dataset
+    train_dataset = PPIDataset(mode='train')
+    valid_dataset = PPIDataset(mode='valid')
+    test_dataset = PPIDataset(mode='test')
+    train_dataloader = GraphDataLoader(train_dataset, batch_size=args.batch_size)
+    valid_dataloader = GraphDataLoader(valid_dataset, batch_size=args.batch_size)
+    test_dataloader = GraphDataLoader(test_dataset, batch_size=args.batch_size)
+
+    # check cuda
+    if args.gpu >= 0 and th.cuda.is_available():
+        device = 'cuda:{}'.format(args.gpu)
+    else:
+        device = 'cpu'
+
+    num_classes = train_dataset.num_labels
+
+    # Extract node features
+    graph = train_dataset[0]
+    feat = graph.ndata['feat']
+
+    # Step 2: Create model =================================================================== #
+    if args.lazy:
+        model = GeniePathLazy(in_dim=feat.shape[-1],
+                              out_dim=num_classes,
+                              hid_dim=args.hid_dim,
+                              num_layers=args.num_layers,
+                              num_heads=args.num_heads,
+                              residual=args.residual)
+    else:
+        model = GeniePath(in_dim=feat.shape[-1],
+                          out_dim=num_classes,
+                          hid_dim=args.hid_dim,
+                          num_layers=args.num_layers,
+                          num_heads=args.num_heads,
+                          residual=args.residual)
+
+    model = model.to(device)
+
+    # Step 3: Create training components ===================================================== #
+    loss_fn = th.nn.BCEWithLogitsLoss()
+    optimizer = optim.Adam(model.parameters(), lr=args.lr)
+
+    # Step 4: training epochs =============================================================== #
+    for epoch in range(args.max_epoch):
+        model.train()
+        tr_loss = 0
+        tr_f1 = 0
+        num_blocks = 0
+        for subgraph in train_dataloader:
+            subgraph = subgraph.to(device)
+            label = subgraph.ndata['label'].to(device)
+            feat = subgraph.ndata['feat']
```

Comment:
Do we need to send the feat to device?

Reply:
It's not necessary, cuz the subgraph is already on the device. Same as the line 84. to(device) deleted in line 84.

Reply:
ok